### PR TITLE
Change a way of eliding filenames

### PR DESCRIPTION
**Pull request by lokson**
Elide hash in filename even if the file length doesn't extend maximum.
Always present file extension in elided form.
Show full url in the title attribute (it will be seen on mouse hover).

Closes: https://github.com/GoogleChrome/lighthouse/issues/1372

### CODE REVIEW

**Comment by lokson** [comment]
Just realized that pretty output needs some changes too. Added WIP status to PR's title

File: lighthouse-core/test/formatter/critical-request-chains-test.js [hunk 1]
```diff
@@ -129,3 +129,32 @@ describe('CRC Formatter', () => {
     assert.ok(truncatedURLRegExp.test(output));
   });
 });
+
+
+describe('parseURL', () => {
+  it('Elides hashes', () => {
+    const url = 'http://example.com/file-f303dec6eec305a4fab8025577db3c2feb418148ac75ba378281399fb1ba670b.css';
+    const result = criticalRequestChainFormatter.parseURL(url);
+    assert.equal(result.file, '/file-f303dec\u2026.css');
```
Comment (ebidel) [inline lighthouse-core/test/formatter/critical-request-chains-test.js]: can you add a test for a hash in the middle:

```
url = 'http://example.com/file-f303dec6eec305a4fab80378281399fb1ba670b-somethingmore.css';
result = criticalRequestChainFormatter.parseURL(url);
assert.equal(result.file, '/file-f303dec\u2026-somethingmore.css');
```

File: lighthouse-core/formatters/critical-request-chains.js [hunk 2]
```diff
@@ -227,8 +227,15 @@ class CriticalRequestChains extends Formatter {
         // And grab the last two parts.
         .split('/').slice(-2).join('/');
 
+    // Elide hash in a filename
+    file = file.replace(/([a-f0-9]{7})[a-f0-9]{13}[a-f0-9]*/g, '$1\u2026');
+    // Elide too long filenames
     if (file.length > MAX_FILENAME_LENGTH) {
-      file = file.slice(0, MAX_FILENAME_LENGTH) + '...';
+      const dotIndex = file.lastIndexOf('.');
+      file =
```
Comment (lokson) [inline lighthouse-core/formatters/critical-request-chains.js]: done

File: lighthouse-core/test/formatter/critical-request-chains-test.js [hunk 3]
```diff
@@ -129,3 +129,32 @@ describe('CRC Formatter', () => {
     assert.ok(truncatedURLRegExp.test(output));
   });
 });
+
+
+describe('parseURL', () => {
```
Comment (lokson) [inline lighthouse-core/test/formatter/critical-request-chains-test.js]: right, added

File: lighthouse-core/test/formatter/critical-request-chains-test.js (hunk 1 above)
Comment (lokson) [inline lighthouse-core/test/formatter/critical-request-chains-test.js]: sure

File: lighthouse-core/test/formatter/critical-request-chains-test.js [hunk 4]
```diff
@@ -129,3 +129,32 @@ describe('CRC Formatter', () => {
...
+  });
+
+  it('Elides long names', () => {
+    const result = criticalRequestChainFormatter.parseURL(superLongName);
+    const expected = '/thisIsASuperLongURLThatWillTriggerFilenameTruncationWhichWe\u2026.js';
+    assert.equal(result.file, expected);
+  });
+
+  it('Elides long names with hash', () => {
+    const url = superLongName.slice(0, -3)
```
Comment (lokson) [inline lighthouse-core/test/formatter/critical-request-chains-test.js]: I'm afraid with indent and const keyword that would exceed lint line width

**Comment by lokson** [comment]
@ebidel, thanks for a rapid fast code review!

**Comment by ebidel** [comment]
Thanks @lokson. Good by me. @paulirish is this what you had in mind?

**Comment by lokson** [comment]
You're welcome!

**Comment by paulirish** [comment]
@lokson looks good. It probably makes sense to move this logic into `url-shim.js`. Wanna do that in a followup PR or this one?
up to you, i'm good with either way.

**Comment by brendankenny** [comment]
> It probably makes sense to move this logic into `url-shim.js`

how much of it though? Just eliding long filenames or also the part about stripping down to the end of the `parsedResourceURL.pathname`?

**Comment by paulirish** [comment]
>> It probably makes sense to move this logic into `url-shim.js`

> how much of it though? Just eliding long filenames or also the part about stripping down to the end of the `parsedResourceURL.pathname`?

both. in devtools source we call it `getDisplayName`.. basically the UI friendly representation of the URL

**Comment by wardpeet** [comment]
maybe for lighthouse a good name would be getShortname?

**Comment by lokson** [comment]
Static `URL.getDisplayName(url)` extracted. Function expects a string and returns a string. This version should be easiest to reuse which comes at a cost of one extra parse call `new URL(url)`. Because not many urls are presented, this is probably irrelevant.

**Comment by paulirish** [comment]
yup! sgtm. nice work @lokson

**Pull request merged**